pkg/auth/providers/oidc/oidc_client.go

```diff
@@ -0,0 +1,210 @@
+	ScopesSupported        []string `json:"scopes_supported"`
+}
+
+//Client implements a httpclient for oidc auth
+type Client struct {
+	httpClient *http.Client
+}
+
+func (o *Client) getUserInfo(authToken string, config *v32.OIDCConfig, pool *x509.CertPool) (*UserInfo, error) {
+	var resp *http.Response
+
+	if config.UserInfoEndpoint == "" {
+		return nil, fmt.Errorf("[generic oidc]: user info enpoint was not provided")
+	}
+	req, err := http.NewRequest("GET", config.UserInfoEndpoint, nil)
+	if err != nil {
+		return nil, fmt.Errorf("[generic oidc]: issue creating GET request: %v", err)
+	}
+	req.Header.Add("Authorization", "Bearer "+authToken)
+	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+
+	if pool != nil {
+		tr := &http.Transport{
+			TLSClientConfig: &tls.Config{
+				RootCAs:            pool,
+				InsecureSkipVerify: true,
+			},
+		}
+		resp, err = tr.RoundTrip(req)
+	} else {
+		resp, err = o.httpClient.Do(req)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("%s: %s", resp.Status, body)
+	}
+	var userInfo UserInfo
+	if err := json.Unmarshal(body, &userInfo); err != nil {
+		return nil, fmt.Errorf("[generic oidc]: failed to decode userinfo: %v", err)
+	}
+	return &UserInfo{
+		Subject:       userInfo.Subject,
+		Profile:       userInfo.Profile,
+		Email:         userInfo.Email,
+		EmailVerified: bool(userInfo.EmailVerified),
+		claims:        body,
+	}, nil
+}
+
+//Authentication Request to Authorization Server - Exchanges Code for Auth Token
+func (o *Client) getAccessTokens(code string, config *v32.OIDCConfig, pool *x509.CertPool) (*Tokens, error) {
+	var resp *http.Response
+
+	form := url.Values{}
+	form.Add("client_id", config.ClientID)
+	form.Add("client_secret", config.ClientSecret)
+	form.Add("response_type", config.ResponseType)
+	form.Add("grant_type", config.GrantType)
+	form.Add("code", code)
+	form.Add("redirect_uri", config.RancherURL)
+
+	req, err := http.NewRequest("POST", config.TokenEndpoint, strings.NewReader(form.Encode()))
+	if err != nil {
+		logrus.Error(err)
+	}
+	req.PostForm = form
+	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+
+	if pool != nil {
+		tr := &http.Transport{
+			TLSClientConfig: &tls.Config{
+				RootCAs:            pool,
+				InsecureSkipVerify: true,
+			},
+		}
+		resp, err = tr.RoundTrip(req)
+	} else {
+		resp, err = o.httpClient.Do(req)
+	}
+
+	if err != nil {
+		logrus.Errorf("[generic oidc] received error from provider: %v", err)
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("[generic oidc] unable to read response body: %v", err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("[generic oidc] response Status: %s. Body: %s", resp.Status, body)
+	}
+
```

Comment:
The code line # 127 to line # 152 can be moved to a method and called from all the above functions that make a request and unmarshal the response body